### PR TITLE
extension: added generic proxy network filter

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: extension: added generic proxy framework
Additional Description:

This is very long story. Please check #18209 for more detailed discussions and background. 
And there is the design proposol https://docs.google.com/document/d/1y_ALyjGp_H1_TRrrRxZxnqENISqeZI6eLn71FAFsZPk/edit#heading=h.a0godkb0g8bc
for why we want the extension.

Due to my personal reasons, I consistently did not finish the final testing of this extension, so it was delayed until now.
I picked this up finally and will complete it in this time.

It's sad the @rojkov cannot sponsor it. So I will self sponsor this extension.

Risk Level: Low. New extension.
Testing: Add.
Docs Changes: Wait.
Release Notes: Wait.
Platform Specific Features: Wait.

### CODE REVIEW

**Comment by wbpcode** [comment]
- [ ] integration test.
- [ ] docs & release note.

File: CODEOWNERS [hunk 1]
```diff
@@ -265,6 +265,8 @@ extensions/filters/http/oauth2 @derekargueta @snowp
 /*/extensions/filters/http/header_to_metadata @zuercher @JuniorHsu
 # zookeeper
 /*/extensions/filters/network/zookeeper_proxy @snowp @JuniorHsu
+# Mata protocol filter
```
Comment (lizan) [inline CODEOWNERS:268]:
```suggestion
# Meta protocol filter
```

**Review by lizan — COMMENTED** [review]
/lgtm api

**Comment by wbpcode** [comment]
cc @mattklein123 Hi, this PR is ready for review. Could you take a look when you have free time? Thanks.

**Comment by wbpcode** [comment]
Because it's long time after last dicussion. Here is some simple context about the extention.

### Why we need this extensions?

Third-party RPC or private protocol support is a common requirement for users of Envoy/Mesh. However it's very complex to developed a new Envoy extension to proxy a third-party RPC or private protocol for now.
This generic proxy is L4 network filter which will provides common route, L7 filter chain, stream management, etc. The developers could add new protocol support to the Envoy by just implementing the core codec of this specific protocol. This reduce much of the complexity of new protocol support.

### About this extension

The design of this extension is come from many others contribution. The initial idea of implement generic proxy in the Envoy is come from @zhaohuabing (aeraki-mesh). The xds matcher based route table is come from @htuch. The async codec API is come from @rojkov. And now, we finally have a implementation for the Envoy community. (The aeraki-mesh will keep their extension because by this way they can add new features more quickly and freely. So we renamed this extension to simple `generic_proxy` to distinguish this extension from the aeraki-mesh's extension and avoid the name conflict in the aeraki-mesh's building.)

### Next step of this extension?

1. Because the async codec API and xds matcher based route table, implementation in this PR is almost complete different with our internal implementation or aeraki-mesh's implementation. So more tests are necessary to ensure everything is work. (The current coverage rate is 95%.)
2. Dubbo codec support. Dubbo proxy has very simple stream model and could be easily to migrate it to this new extension. After that, we can avoid the cost of maintaining the dubbo proxy. Of course, we will have a long, long time to gradually deprecate the old dubbo proxy.
3. More features, such rds support, observability, route level config, etc.

**Comment by mattklein123** [comment]
@wbpcode this is too much code for me to reasonably review. I'm not sure the best way of moving this forward. I think rewriting Dubbo to use this would give me a lot more confidence. How would you suggest incrementally moving this forward?

**Comment by wbpcode** [comment]
@mattklein123 As I originally planned, after this PR is merged, I would first move dubbo decoder and encoder to extensions/common and then implement dubbo codec for generic proxy based on these encoder/decoder.

**Comment by wbpcode** [comment]
Yeah, this is a super big PR that hard to review. 🤣 But this extension still be marked as WIP, so we still have chance to correct it in the future if there are something wrong.

IMO, the core should be the message API and codec API, you can check these code in the `genetic_proxy/interface`.

**Comment by wbpcode** [comment]
This extension was moved to contrib as our offline discussion. @mattklein123

**Review by mattklein123 — APPROVED** [review]
Thanks. I'm OK merging this into contrib. Once Dubbo has been rewritten to use this and vetted in production I think we can deprecate/remove the existing Dubbo filter and move this all into core. Thank you!

**Pull request merged**